otter/convergence/service.py

```diff
@@ -91,6 +91,38 @@ def update_group_state(group, old_state):
 
 
 @do
+def _execute_steps(steps):
```

Comment:
Thank you for splitting this out :) I was going to suggest a separate refactor PR before even looking at this one.